Bug Fix for Issue #1445 (Anchor parent directive always renders at the time of the child)

shreyasyellenki opened the pull request:
The Anchor parent directive was always rendering same time of the child because the calculation of the `start_time_ms` attribute for `ActivityDirective` was incorrect in `getActivityDirectiveStartTimeMs.` The previous algorithm relied on nested casting and had a difficult flow to follow. The proposed change finds the anchor parent directive's `start_time_ms`, then handles if the child is anchored to the start or end. Previously, `ActivityDirectiveForm`  compared a directive's `start_offset` to the plan's start time to display Absolute Start Time. This ignores any anchor relationships leading to incorrect Absolute Start Times being displayed, instead it should be based off of `start_time_ms`.

shreyasyellenki committed:
fix: change how Activity Directive start_time_ms is calculated

Incorrect Activity Directive start_time_ms when anchors were involved.
This results in Activity Directives displaying in different locations than their respective span.
Remove ActivityDirectiveForm reliance on start_offset to display Absolute Start Time
as anchors are not considered in that situation.
Modify getActivityDirectiveStartTimeMs in src/utilities/time.ts.
Calculate start_time_ms without nested casting and handle anchored_to_start in separate chunk.
Make ActivityDirectiveForm use start_time_ms to display correct Absolute Start Time.

duranb commented on src/components/activity/ActivityDirectiveForm.svelte:
```diff
@@ -101,11 +101,12 @@
     (activityTypes ?? []).find(({ name: activityTypeName }) => activityDirective?.type === activityTypeName) ?? null;
   $: {
-    const startTimeMs = getUnixEpochTimeFromInterval(
+    /* const startTimeMs = getUnixEpochTimeFromInterval(
```
Do you mind clearing out any commented out code, please?

duranb commented on src/components/activity/ActivityDirectiveForm.svelte:
```diff
@@ -101,11 +101,12 @@
-    );
-    startTime = formatDate(new Date(startTimeMs), $plugins.time.primary.format);
+    );*/
+    const startTimeMs = activityDirective.start_time_ms;
```
This probably also needs to take into account if `revision` exists like it currently does, right?

shreyasyellenki replied:
Changed: Previously, Absolute Start Time for revisions was based only on `start_offset` and did not account for anchors. Relabeled the existing schema for `ActivityDirectiveRevision` as `ActivityDirectiveDB,` and then extended with a "new"  type called `ActivityDirective` which has an extra attribute of `start_time_ms`. Calculate the `start_time_ms` for revisions when they are retrieved from the database. Calculation for `start_time_ms` and consequently Absolute Start Time is not possible if a revision state leads to a current anchor cycle. The user is notified of this and the Absolute Start Time field is left empty.

shreyasyellenki committed:
refactor: absolute start time for revisions

Rename original ActivityDirectiveRevision to ActivityDirectiveRevisionDB, and then extend to a new type called ActivityDirectiveRevision which has a start_time_ms parameter similar to ActivityDirective. Create function that calculates absolute start time for revisions.

shreyasyellenki committed:
test: addAbsoluteTimeToRevision, refactor: Anchor cycles

Add unit test for addAbsoluteTimeToRevision

Caught error for when past ActivityDirective revisions lead to current anchor cycles and show handle that absolute start time cannot be calculated.

shreyasyellenki committed:
refactor: anchor cycle warning

Instead of a placeholder absolute start time of all 0's, the Absolute start time field will now be empty if a revision leads to a present day anchor cycle

shreyasyellenki commented:
The Quality Gate message "failed for duplicate code", is because I copied over a plan from another test and used it in the unit test for `addAbsoluteTimeToRevision`

shreyasyellenki committed:
Wiped node modules and re-installed

Ran into some node.js issues when trying to start dev server, wiped existing node modules and svelte-kit and re-installed.
